### ISSUE TITLE
Lock file error when hosted on Azure App service

### ISSUE THREAD

**Issue by ShemJM** [elsa-workflows/elsa-core#4796]
Hi, I have added Elsa to my existing asp.net core api and published it to a Linux Azure App Service. It works locally but now it is published I am getting the following error

RequestPath
/elsa/api/workflow-definitions
System.InvalidOperationException: Failed to ensure that lock file directory /home/site/wwwroot/App_Data/locks exists
 ---> System.IO.IOException: Read-only file system : '/home/site/wwwroot/App_Data'
 
 I can't seem to find anything in the source code that is relevant, other than the public class DefaultWorkflowRuntime : IWorkflowRuntime
 
 Not sure what I can do here to get this to work

**Comment by sfmskywalker** [comment]
This issues stems from the DistributedLock package, which is a distributed lock provider that is configured to use the local file system as a shared resource by default. From the error it seems that your application doesn't have write permission to the `/home/site/wwwroot/App_Data`. By default, the file system distributed lock provider is configured to store locks in `App_Data/locks`.

One thing you could try easily is to create the App_Data/locks directory manually and include a simple txt file in the locks directory to ensure it exists in your Azure app Service instance.

If it still fails to create lock files then you need to somehow grant write permissions to the App_Data/locks directory. I don't know off-hand how you would go about that, but you should be able your google your way into that.

**Comment by ShemJM** [comment]
Thanks, failing that. Are there any options to configure this within Elsa?

**Comment by sfmskywalker** [comment]
Yes, Elsa exposes the configuration of the DistributedLock service directly to your program, which means you could use a different folder path to which your application does have write permissions. And, even better; you can use a different provider, such as a Database, Redis or Azure Storage provider, or any of their [supported providers](https://github.com/madelson/DistributedLock?tab=readme-ov-file#implementations).

First make sure that you have the desired package installed (e.g. [distributedLock.SqlServer](https://github.com/madelson/DistributedLock/blob/master/docs/DistributedLock.SqlServer.md) in case you want to use SqlServer as the provider).

Then tell Elsa to use the appropriate provider. For example, the following uses the built-in FileSystem provider using a specific path:

```csharp
services.AddElsa(elsa =>
{
   elsa.UseRuntime(runtime => 
   {
      runtime.DistributedLockProvider = _ => new FileDistributedSynchronizationProvider(new DirectoryInfo(Path.Combine(Environment.CurrentDirectory, "//some//path//to//locks")));
   };
};
```

To use the SqlServer provider, replace the assignment to `runtime.DistributedLockProvider` with this:

```csharp
runtime.DistributedLockProvider = _ => new SqlDistributedSynchronizationProvider(sqlServerConnectionString!);
```

**Comment by avhovsepyan** [comment]
Hi. I tried and found that for me works UseWorkflowRuntime(), instead of UseRuntime().
Instead of this:
`services.AddElsa(elsa =>
{
   elsa.UseRuntime(runtime => 
   {
      runtime.DistributedLockProvider = _ => new FileDistributedSynchronizationProvider(new DirectoryInfo(Path.Combine(Environment.CurrentDirectory, "//some//path//to//locks")));
   };
};`

Use this:
`services.AddElsa(elsa =>
{
   elsa.UseWorkflowRuntime(runtime => 
   {
      runtime.DistributedLockProvider = _ => new FileDistributedSynchronizationProvider(new DirectoryInfo(Path.Combine(Environment.CurrentDirectory, "//some//path//to//locks")));
   };
};`

**Comment by ShemJM** [comment]
Yes, thanks. I have done the same.
